File: profile/bench/test_sieve.py
import array
import math

def sieve_array(n : int) -> list[int]:
    '''Return a list of prime numbers between 2 and n.'''
    arr: array = array.array('b', (1 for _ in range(n+1)))

    for i in range(2, math.ceil(math.sqrt(n)) + 1):
        if arr[i]:
            for j in range(i*i, n+1, i):
                arr[j] = 0

    return list(i for i, is_prime in enumerate(arr[2:n+1], 2) if is_prime)


def sieve_bytearray(n : int) -> list[int]:
    '''Return a list of prime numbers between 2 and n.'''
    arr: bytearray = bytearray(b'\x01'*(n+1))

    for i in range(2, math.ceil(math.sqrt(n)) + 1):
        if arr[i]:
            for j in range(i*i, n+1, i):
                arr[j] = 0

    return list(i for i, is_prime in enumerate(arr[2:n+1], 2) if is_prime)


def sieve(n: int) -> list[int]:
    '''Return a list of prime numbers between 2 and n.'''
    arr: list[bool] = [True] * (n+1)

    for i in range(2, math.ceil(math.sqrt(n)) + 1):
        if arr[i]:
            for j in range(i*i, n+1, i):
                arr[j] = False

    return list(i for i, is_prime in enumerate(arr[2:n+1], 2) if is_prime)


def test_sieve_array(benchmark):
    '''Benchmark the array-based sieve implementation.'''
    result = benchmark(sieve_array, 10000)
    assert result[-1] == 9973


def test_sieve_bytearray(benchmark):
    '''Benchmark the bytearray-based sieve implementation.'''
    result = benchmark(sieve_bytearray, 10000)
    assert result[-1] == 9973


def test_sieve_list(benchmark):
    '''Benchmark the list-based sieve implementation.'''
    result = benchmark(sieve, 10000)
    assert result[-1] == 9973

File: profile/bench/test_sieve.cpp
#include "benchmark/benchmark.h"

#include <array>
#include <cmath>
#include <memory>
#include <unordered_map>
#include <vector>


std::vector<long> sieve(long n) {
    std::vector<bool> arr((unsigned long)n, true);

    for (long i = 2; i <= (long)std::sqrt(n); ++i) {
        if (arr[i]) {
            for (long j = i*i; j < n; j += i) {
                arr[j] = false;
            }
        }
    }

    std::vector<long> result;
    for (long k = 2; k < n; ++k) {
        if (arr[k]) {
            result.emplace_back(k);
        }
    }
    return result;
}


std::vector<long> sieve_array(long n) {
    std::unique_ptr<bool[]> arr = std::make_unique<bool[]>(n+1);

    for (long i = 2; i <= n; ++i) {
        arr[i] = true;
    }

    for (long j = 2; j <= (long)std::sqrt(n); ++j) {
        if (arr[j]) {
            for (long k = j*j; k < n; k += j) {
                arr[k] = false;
            }
        }
    }

    std::vector<long> result;
    for (long l = 2; l < n; ++l) {
        if (arr[l]) {
            result.emplace_back(l);
        }
    }
    return result;
}


std::vector<long> sieve_map(long n) {
    std::unordered_map<long, bool> arr;

    for (long i = 2; i <= n; ++i) {
        arr[i] = true;
    }

    for (long j = 2; j <= (long)std::sqrt(n); ++j) {
        if (arr[j]) {
            for (long k = j*j; k < n; k += j) {
                arr[k] = false;
            }
        }
    }

    std::vector<long> result;
    for (long l = 2; l < n; ++l) {
        if (arr[l]) {
            result.emplace_back(l);
        }
    }
    return result;
}


static void BM_Sieve_Vector(benchmark::State& state) {
    for (auto _ : state) {
        benchmark::DoNotOptimize(sieve(100000));
    }
}
BENCHMARK(BM_Sieve_Vector);


static void BM_Sieve_Array(benchmark::State& state) {
    for (auto _ : state) {
        benchmark::DoNotOptimize(sieve_array(100000));
    }
}
BENCHMARK(BM_Sieve_Array);


static void BM_Sieve_Map(benchmark::State& state) {
    for (auto _ : state) {
        benchmark::DoNotOptimize(sieve_map(100000));
    }
}
BENCHMARK(BM_Sieve_Map);


BENCHMARK_MAIN();
